src/Microsoft.OData.Core/JsonLight/ODataJsonLightPropertyAndValueDeserializer.cs

```diff
@@ -1059,154 +1163,1458 @@ protected EdmTypeKind GetNonEntityValueKind()
+        /// </summary>
+        /// <param name="expectedPropertyTypeReference">The expected type reference of the property to read.</param>
+        /// <param name="propertyAndAnnotationCollector">The duplicate property names checker to use.</param>
+        /// <returns>
+        /// A task that represents the asynchronous read operation.
+        /// The value of the TResult parameter contains an <see cref="ODataProperty"/> representing the read property.
+        /// </returns>
+        /// <remarks>
+        /// The method assumes that the ReadPayloadStartAsync has already been called and it will not call ReadPayloadEndAsync.
+        /// </remarks>
+        private async Task<ODataProperty> ReadTopLevelPropertyImplementationAsync(
+            IEdmTypeReference expectedPropertyTypeReference,
+            PropertyAndAnnotationCollector propertyAndAnnotationCollector)
+        {
+            Debug.Assert(
+                expectedPropertyTypeReference == null || !expectedPropertyTypeReference.IsODataEntityTypeKind(),
+                "If the expected type is specified it must not be an entity type.");
+            Debug.Assert(propertyAndAnnotationCollector != null, "propertyAndAnnotationCollector != null");
+
+            expectedPropertyTypeReference = this.UpdateExpectedTypeBasedOnContextUri(expectedPropertyTypeReference);
+
+            object propertyValue = missingPropertyValue;
+            Collection<ODataInstanceAnnotation> customInstanceAnnotations = new Collection<ODataInstanceAnnotation>();
+
+            // Check for the special top-level OData 3.0 null marker in order to accommodate
+            // the responses written by 6.x version of this library.
+            if (await this.IsTopLevel6xNullValueAsync().ConfigureAwait(false))
+            {
+                // NOTE: When reading a null value we will never ask the type resolver (if present) to resolve the
+                //       type; we always fall back to the expected type.
+                this.ReaderValidator.ValidateNullValue(
+                    expectedPropertyTypeReference,
+                    validateNullValue: true,
+                    propertyName: null,
+                    isDynamicProperty: null);
+
+                // We don't allow properties or non-custom annotations in the null payload.
+                await this.ValidateNoPropertyInNullPayloadAsync(propertyAndAnnotationCollector)
+                    .ConfigureAwait(false);
+
+                propertyValue = null;
+            }
+            else
+            {
+                string payloadTypeName = null;
+                Tuple<bool, string> readingResourcePropertyResult = await this.ReadingResourcePropertyAsync(
+                    propertyAndAnnotationCollector,
+                    expectedPropertyTypeReference).ConfigureAwait(false);
+                payloadTypeName = readingResourcePropertyResult.Item2;
+
+                if (readingResourcePropertyResult.Item1)

```

Comment:
Maybe you could store the `.Item1` value in a descriptive variable and execute the if statement against the variable like you have done other places for better readability.

Reply:
Fixed